### PR TITLE
Payment driver shutdown

### PR DESCRIPTION
**Pull request by Wiezzel**
When Ctrl+c is hit let the payment service shut down gracefully and send out all pending transactions:
 * Stop accepting calls to `SchedulePayment` and `ValidateAllocation`.
 * Trigger tx sending routing once for all drivers.
 * Start calling tx confirmation routine in a loop.
 * Stop after 10 seconds.
 * Stop immediately if Ctrl+c is hit again.
 
 ---
##### Scenario A
1. Run `payment_api` + `invoice_flow` using `zksync` driver.
2. When `invoice_flow` displays `Waiting for payment...` go quickly to the `payment_api` console and hit Ctrl+c.
3. `invoice_flow` **should fail**.
4. `payment_api` should stop after 10 seconds.
5. Go to zkscan and check if transaction has been made nonetheless.

##### Scenario B
1. Run `payment_api` + `invoice_flow` using `zksync` driver.
2. When `invoice_flow` displays `Waiting for payment...` go quickly to the `payment_api` console and hit Ctrl+c **twice**.
3. `invoice_flow` **should fail**.
4. `payment_api` should stop immediately.
5. Go to zkscan and check if transaction has **not** been made.PAYMENT_SHUTDOWN_TIMEOUT_SECS

##### Scenario C
Same as A but use `erc20` driver and set `PAYMENT_SHUTDOWN_TIMEOUT_SECS` to `20` or `30` (because erc-20 is slower).

##### Scenario D
Same as B but use `erc20` driver.

##### Scenarios X, Y, Z
Try single and double Ctrl+c with real yagna daemon with both drivers when runnning a yapapi/yajsapi task.

### CODE REVIEW

File: core/payment/src/processor.rs [hunk 1]
```diff
@@ -585,4 +599,29 @@ impl PaymentProcessor {
         let result = driver_endpoint(&driver).send(msg).await??;
         Ok(result)
     }
+
+    pub async fn shut_down(&mut self, timeout: Duration) {
+        self.in_shutdown = true;
```
Comment (maaktweluit) [inline core/payment/src/processor.rs:604]: naming is not consistent: `shutdown` vs `shut_dowm`.. I like the first one more

**Review by maaktweluit — COMMENTED** [review]
LGTM!

Some nitty comments / questions...

Will test and approve when you un-draft the PR

File: core/payment-driver/base/src/cron.rs [hunk 2]
```diff
@@ -18,57 +17,31 @@ pub use async_trait::async_trait;
...
-                    }
-                }
-            });
+            let driver = act.driver.clone();
+            Arbiter::spawn(async move { driver.confirm_payments().await });
         });
     }
 
-    fn start_payment_job(&mut self, ctx: &mut Context<Self>) {
+    fn start_sendout_job(&mut self, ctx: &mut Context<Self>) {
```
[the remark this answers is not in the document]
Comment (Wiezzel) [reply core/payment-driver/base/src/cron.rs:41]: 'Send out' is a verb, 'sendout' or 'send-out' is a noun. Compare 'break down' vs 'breakdown'.

File: core/payment-driver/erc20/src/driver.rs [hunk 3]
```diff
@@ -4,6 +4,8 @@
     Please limit the logic in this file, use local mods to handle the calls.
 */
 // Extrnal crates
+use chrono::{Duration, Utc};
+use futures::lock::Mutex;
```
[the remark this answers is not in the document]
Comment (Wiezzel) [reply core/payment-driver/erc20/src/driver.rs:8]: I just go with this one as default choice. Are there any downsides of doing this?

File: core/payment/src/processor.rs (hunk 1 above)
Comment (Wiezzel) [reply core/payment/src/processor.rs:604]: Same issue as 'sendout' vs 'send out'.

**Review by maaktweluit — APPROVED** [review]
LGTM! Tested OK :)

**Comment by Wiezzel** [comment]
@maaktweluit A significant improvement has been made: https://github.com/golemfactory/yagna/pull/1181/commits/7a5235ca3edcdd4850d6f748046eb379caca0d87

---
Testing instructions:
1. Run the original test but after stopping `payment_api` also check in DB that `payment` entry has been created and that invoice record *for requestor* has been marked as `SETTLED` (not for provider – we do not want to wait for remote host during shutdown).
2. Run `payment_api` and/or yagna service and stop it immediately after it's ready without making any payment operations. It should shut down immediately instead of spending 10s waiting in vain.

**Review by maaktweluit — APPROVED** [review]
Nice updates! 

Tested OK :)

**Review by maaktweluit — APPROVED** [review]

**Pull request merged**